pkgs/cupertino_http/lib/src/cupertino_api.dart

```diff
@@ -1111,279 +793,195 @@ class MutableURLRequest extends URLRequest {
+          onComplete(
+              URLSession._(nsSession,
+                  isBackground: isBackground, hasDelegate: true),
+              URLSessionTask._(nsTask),
+              nsError);
+        }
+      },
+    );
+
+    if (onRedirect != null) {
+      ncb.NSURLSessionDataDelegate.addToBuilderAsListener(protoBuilder,
+          // ignore: lines_longer_than_80_chars
+          URLSession_task_willPerformHTTPRedirection_newRequest_completionHandler_:
+              (nsSession, nsTask, nsResponse, nsRequest, nsRequestCompleter) {
+        final request = URLRequest._(nsRequest);
+        URLRequest? redirectRequest;
+
+        try {
+          final response =
+              URLResponse._exactURLResponseType(nsResponse) as HTTPURLResponse;
+          redirectRequest = onRedirect(
+              URLSession._(nsSession,
+                  isBackground: isBackground, hasDelegate: true),
+              URLSessionTask._(nsTask),
+              response,
+              request);
+          nsRequestCompleter.call(redirectRequest?._nsObject);
+        } catch (e) {
+          // TODO(https://github.com/dart-lang/ffigen/issues/386): Package
+          // this exception as an `Error` and call the completion function
+          // with it.
+        }
+      });
+    }
+
+    if (onResponse != null) {
+      ncb.NSURLSessionDataDelegate.addToBuilderAsListener(protoBuilder,
+          URLSession_dataTask_didReceiveResponse_completionHandler_:
+              (nsSession, nsDataTask, nsResponse, nsCompletionHandler) {
+        final exactResponse = URLResponse._exactURLResponseType(nsResponse);
+        final disposition = onResponse(
+            URLSession._(nsSession,
+                isBackground: isBackground, hasDelegate: true),
+            URLSessionTask._(nsDataTask),
+            exactResponse);
+        nsCompletionHandler.call(disposition);
+      });
+    }
+
+    if (onData != null) {
+      ncb.NSURLSessionDataDelegate.addToBuilderAsListener(protoBuilder,
+          URLSession_dataTask_didReceiveData_: (nsSession, nsDataTask, nsData) {
+        onData(
+            URLSession._(nsSession,
+                isBackground: isBackground, hasDelegate: true),
+            URLSessionTask._(nsDataTask),
+            nsData);
+      });
+    }
+
+    if (onFinishedDownloading != null) {
+      final asyncFinishDownloading =
+          // ignore: lines_longer_than_80_chars
+          ncb.ObjCBlock_ffiVoid_NSCondition_NSURLSession_NSURLSessionDownloadTask_NSURL
+              .listener((nsCondition, nsSession, nsTask, nsUrl) {
+        try {
+          onFinishedDownloading(
+              URLSession._(nsSession,
+                  isBackground: isBackground, hasDelegate: true),
+              URLSessionDownloadTask._(nsTask),
+              nsurlToUri(nsUrl));
+        } finally {
+          nsCondition.unlock();
+        }
+      });
+
+      final downloadDelegate = objc.getProtocol('NSURLSessionDownloadDelegate');
```

Comment:
I've filed https://github.com/dart-lang/native/issues/1675 so you won't have to do this in future.